### PR TITLE
Miscellaneous fixes

### PR DESCRIPTION
**Pull request by dimbleby**
A small collection of miscellaneous fixes, none of which have individually seemed to me worth the bother of a pull request.  But I've now got enough of them queued up that, well, here we are.

Hopefully the commits and their messages are self-explanatory, these are all intended as uncontroversial.  But if some of them are for some reason harder to merge than others then say so and this can be broken up into more PRs.

### CODE REVIEW

File: src/poetry/repositories/abstract_repository.py [hunk 1]
```diff
@@ -30,7 +31,7 @@ def search(self, query: str) -> list[Package]:
     @abstractmethod
     def package(
         self,
-        name: str,
+        name: NormalizedName,
```
Comment (radoering) [inline src/poetry/repositories/abstract_repository.py:34]: I'm wondering that mypy doesn't complain. There are more subclasses that still take pretty name and iirc some need the pretty name.

File: src/poetry/puzzle/provider.py [hunk 2]
```diff
@@ -572,7 +572,7 @@ def complete_package(
                 dependency_package = DependencyPackage(
                     dependency,
                     self._pool.package(
-                        package.pretty_name,
+                        package.name,
```
Comment (radoering) [inline src/poetry/puzzle/provider.py:575]: Obviously, there is no test, but I expect this to cause some name changes in the lock file (again). (See other comment.) I'd like to avoid that if there is no good reason.

File: src/poetry/repositories/abstract_repository.py (hunk 1 above)
Comment (dimbleby) [reply src/poetry/repositories/abstract_repository.py:34]: Could you point out what you mean?  Like mypy, I didn't notice anything (though per the commit it's some weeks since I looked at this and so it's not fresh in my mind).

File: src/poetry/puzzle/provider.py (hunk 2 above)
Comment (dimbleby) [reply src/poetry/puzzle/provider.py:575]: why would this cause a change in the lock file?

File: src/poetry/repositories/abstract_repository.py (hunk 1 above)
Comment (radoering) [reply src/poetry/repositories/abstract_repository.py:34]: https://github.com/python-poetry/poetry/blob/0c8e5b43d30d86d6ba7dc9b3068db26042e9cc53/src/poetry/repositories/cached_repository.py#L70-L78

https://github.com/python-poetry/poetry/blob/0c8e5b43d30d86d6ba7dc9b3068db26042e9cc53/src/poetry/repositories/legacy_repository.py#L48-L72

https://github.com/python-poetry/poetry/blob/0c8e5b43d30d86d6ba7dc9b3068db26042e9cc53/src/poetry/repositories/repository_pool.py#L110-L127

File: src/poetry/puzzle/provider.py (hunk 2 above)
Comment (radoering) [reply src/poetry/puzzle/provider.py:575]: Because the pretty name is "lost" and thus the canonicalized name is written in the lock file.

File: src/poetry/repositories/abstract_repository.py (hunk 1 above)
Comment (dimbleby) [reply src/poetry/repositories/abstract_repository.py:34]: ah, somehow machine and man both missed that the legacy repository uses the un-normalized name: [here](https://github.com/python-poetry/poetry/blob/0c8e5b43d30d86d6ba7dc9b3068db26042e9cc53/src/poetry/repositories/cached_repository.py#L77).

I'll remove this change from the collection

File: src/poetry/puzzle/provider.py [hunk 3]
```diff
@@ -945,19 +945,9 @@ def _merge_dependencies_by_constraint(
...
-                    # No marker or only extras
-                    continue
-
-                new_markers.append(marker)
-
-            if not new_markers:
-                continue
 
+        for constraint, _deps in by_constraint.items():
+            new_markers = [d.marker for d in _deps]
```
Comment (radoering) [inline src/poetry/puzzle/provider.py:950]: Does that maybe even fix #5429? I'd appreciate if you could check and if that's the case split this change into a separate PR and add a test. 🙏 

By the way, line 952 can probably be simplified to `dep.marker = MarkerUnion.of(*new_markers)`. If I don't miss something the marker of the first dependency is considered twice unnecessarily at the moment.

Comment (dimbleby) [reply src/poetry/puzzle/provider.py:950]: I considered MarkerUnion.of() but it's heading in the wrong direction with respect to https://github.com/python-poetry/poetry-core/pull/530, which removes much of the simplification from the `of()` methods

Comment (radoering) [reply src/poetry/puzzle/provider.py:950]: From an API perspective, I wonder if it might make sense to have two versions of `of()` (better names welcome). One that we use internally for cnf/dnf and one (that basically just does the `dnf(cnf())` trick) that can be used in cases like this. It seems more natural to use something like `of()` here. Or maybe just add a parameter to `of()` which is actively set by cnf/dnf?

Comment (dimbleby) [reply src/poetry/puzzle/provider.py:950]: so far as I can see this doesn't touch #5429

https://github.com/python-poetry/poetry-core/pull/530 is more-or-less making `XXXMarker.of()` a private method.  

I'm going to tweak this particular line so that it goes `AnyMarker().union(MarkerUnion(*new_markers)` which is certainly a bit artificial.  But I think it's the most efficient thing to do, or anyway will be if-and-when https://github.com/python-poetry/poetry-core/pull/530 merges.

I wouldn't be opposed to some new API allowing union-of-lots-of-markers to be more conveniently constructed in one go, but also I'm not especially excited about it.

Comment (radoering) [reply src/poetry/puzzle/provider.py:951]: I took a closer look. It's part of the issue. See #7175

Comment (dimbleby) [reply src/poetry/puzzle/provider.py:951]: Oh, how satisfying.  Nice.

File: pyproject.toml [hunk 4]
```diff
@@ -56,7 +56,7 @@ crashtest = "^0.4.1"
 dulwich = "^0.20.46"
 filelock = "^3.8.0"
 html5lib = "^1.0"
-importlib-metadata = { version = "^4.4", python = "<3.10" }
+importlib-metadata = { version = "^5.1", python = "<3.10" }
```
Comment (radoering) [inline pyproject.toml:59]: According to https://pypi.org/project/importlib-metadata/, 4.6 is included in python 3.10. If 4.6 is enough for Python 3.10, why should we require 5.1 for python < 3.10?

Comment (dimbleby) [reply pyproject.toml:59]: no big deal either way, but IMO it's as well for application-type code to stay current. What is gained by allowing old versions of importlib-metadata?

Indeed we're currently explicitly excluding new versions of importlib-metadata, which seems a reasonable thing to happen by accident, but a strange thing to choose deliberately.

Comment (radoering) [reply pyproject.toml:59]: Yeah, `^4.4` isn't good, either. Probably, it should be `>=4.4,<6`.

Re allowing old versions: If you install poetry in its own virtual env (as recommended) it doesn't matter. But then we could also pin dependencies. 🤷

Comment (dimbleby) [reply pyproject.toml:59]: sure, it's a can of worms.  But for applications - and poetry really is intended to be installed in its own virtual environment - I don't see meaningful upside in allowing a broad range.  There certainly is downside: it would be some sort of commitment to support old versions of importlib-metadata (snd there is no testing to verify that this will work).

However I don't think this is important enough that I care to worry about it: if you're asking me to make a change, please say so.

Comment (radoering) [reply pyproject.toml:59]: > There certainly is downside: it would be some sort of commitment to support old versions of importlib-metadata (snd there is no testing to verify that this will work).

Personally, I like to keep constraints as broad as possible and only restrict them if a version is known to be incompatible. (At least, experienced users can easily exclude a bad version by themselves, but cannot allow an excluded version as easy.)

It might be a while before we find the time to discuss our update strategy internally.  So yes, maybe just change it and I'll merge it.

**Review by radoering — APPROVED** [review]
